README.md
![astar-cover](https://user-images.githubusercontent.com/40356749/135799652-175e0d24-1255-4c26-87e8-447b192fd4b2.gif)

<div align="center">

[![Integration Action](https://github.com/AstarNetwork/Astar/workflows/Integration/badge.svg)](https://github.com/AstarNetwork/Astar/actions)
[![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/AstarNetwork/Astar)](https://github.com/AstarNetwork/Astar/tags)
[![Substrate version](https://img.shields.io/badge/Substrate-3.0.0-brightgreen?logo=Parity%20Substrate)](https://substrate.dev/)
[![License](https://img.shields.io/github/license/AstarNetwork/Astar?color=green)](https://github.com/AstarNetwork/Astar/blob/production/shiden/LICENSE)
 <br />
[![Twitter URL](https://img.shields.io/twitter/follow/AstarNetwork?style=social)](https://twitter.com/AstarNetwork)
[![Twitter URL](https://img.shields.io/twitter/follow/ShidenNetwork?style=social)](https://twitter.com/ShidenNetwork)
[![YouTube](https://img.shields.io/youtube/channel/subscribers/UC36JgEF6gqatVSK9xlzzrvQ?style=social)](https://www.youtube.com/channel/UC36JgEF6gqatVSK9xlzzrvQ)
[![Docker](https://img.shields.io/docker/pulls/staketechnologies/astar-collator?logo=docker)](https://hub.docker.com/r/staketechnologies/astar-collator)
[![Discord](https://img.shields.io/badge/Discord-gray?logo=discord)](https://discord.gg/astarnetwork)
[![Telegram](https://img.shields.io/badge/Telegram-gray?logo=telegram)](https://t.me/PlasmOfficial)
[![Medium](https://img.shields.io/badge/Medium-gray?logo=medium)](https://medium.com/astar-network)

</div>

Astar Network is an interoperable blockchain based the Substrate framework and the hub for dApps within the Polkadot Ecosystem.
With Astar Network and Shiden Network, people can stake their tokens to a Smart Contract for rewarding projects that provide value to the network.

For contributing to this project, please read our [Contribution Guideline](./CONTRIBUTING.md).

## Building From Source

> This section assumes that the developer is running on either macOS or Debian-variant operating system. For Windows, although there are ways to run it, we recommend using [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10) or from a virtual machine for stability.

Execute the following command from your terminal to set up the development environment and build the node runtime.

```bash
# install Substrate development environment via the automatic script
$ curl https://getsubstrate.io -sSf | bash -s -- --fast

# clone the Git repository
$ git clone --recurse-submodules https://github.com/AstarNetwork/Astar.git

# change current working directory
$ cd Astar

# compile the node
# note: you may encounter some errors if `wasm32-unknown-unknown` is not installed, or if the toolchain channel is outdated
$ cargo build --release

# show list of available commands
$ ./target/release/astar-collator --help
```

### Building with Nix

```bash
# install Nix package manager:
$ curl https://nixos.org/nix/install | sh

# run from root of the project folder (`Astar/` folder)
$ nix-shell -I nixpkgs=channel:nixos-21.05 third-party/nix/shell.nix --run "cargo build --release"
```

## Running a Collator Node

To set up a collator node, you must have a fully synced node with the proper arguments, which can be done with the following command.

```bash
# start the Shiden collator node with
$ ./target/release/astar-collator \
  --base-path <path to save blocks> \
  --name <node display name> \
  --port 30333 \
  --ws-port 9944 \
  --rpc-port 9933 \
  --telemetry-url 'wss://telemetry.polkadot.io/submit/ 0' \
  --rpc-cors all \
  --collator
```

Now, you can obtain the node's session key by sending the following RPC payload.

```bash
# send `rotate_keys` request
$ curl -H 'Content-Type: application/json' --data '{ "jsonrpc":"2.0", "method":"author_rotateKeys", "id":1 }' localhost:9933

# should return a long string of hex, which is your session key
{"jsonrpc":"2.0","result":"<session key in hex>","id":1}
```

After this step, you should have a validator node online with a session key for your node.
For key management and validator rewards, consult our [validator guide online](https://docs.astar.network/build/validator-guide/configure-node).

## Run RPC Tests

RPC tests suite can be run for any release. To run tests go to https://github.com/AstarNetwork/Astar/actions/workflows/rpcTest.yml. Click Run workflow, in the dropdown input the release version tag you want to run the test suite. Then click the green Run workflow button to start the test suite. 

![Screenshot from 2022-07-07 15-28-46](https://user-images.githubusercontent.com/874046/177785570-330c6613-237d-4190-bfed-69876209daf6.png)

## Workspace Dependency Handling

All dependencies should be listed inside the workspace's root `Cargo.toml` file.
This allows us to easily change version of a crate used by the entire repo by modifying the version in a single place.

Right now, if **non_std** is required, `default-features = false` must be set in the root `Cargo.toml` file (related to this [issue](https://github.com/rust-lang/cargo/pull/11409)). Otherwise, it will have no effect, causing your compilation to fail.
Also `package` imports aren't properly propagated from root to sub-crates, so defining those should be avoided.

Defining _features_ in the root `Cargo.toml` is additive with the features defined in concrete crate's `Cargo.toml`.

**Adding Dependency**
1. Check if the dependency is already defined in the root `Cargo.toml`
    1. if **yes**, nothing to do, just take note of the enabled features
    2. if **no**, add it (make sure to use `default-features = false` if dependency is used in _no_std_ context)
2. Add `new_dependecy = { workspace = true }` to the required crate
3. In case dependency is defined with `default-features = false` but you need it in _std_ context, add `features = ["std"]` to the required crate.
 
## Contributing

We welcome contributions to this repository! To get started, please read the following guidelines carefully.


### Submitting a Pull Request

When submitting a Pull Request (PR), please make sure to include the necessary labels to help us categorize and prioritize your changes. PRs should contain at least one of the following mandatory labels:

- **runtime**: Use this label for changes or additions to the code that relate to runtime **(shiden, astar, shibuya, local)**
- **client**: PRs related to client.
- **ci**: PRs related to CI (workflows).
- **tests**: PRs related to xcm-simulator tests,rpc-tests or any other kind of test.
- **other**: PRs related to third-party, scripts and all other things which don't have a definite label.


By including these labels in your PR submission, you can help us to streamline the development process and ensure that your changes are properly reviewed and merged into the codebase.


## Further Reading

* [Official Documentation](https://docs.astar.network/)
* [Whitepaper](https://github.com/AstarNetwork/plasmdocs/blob/master/wp/en.pdf)
* [Whitepaper(JP)](https://github.com/AstarNetwork/plasmdocs/blob/master/wp/jp.pdf)
* [Subtrate Developer Hub](https://substrate.dev/docs/en/)
* [Substrate Glossary](https://substrate.dev/docs/en/knowledgebase/getting-started/glossary)
* [Substrate Client Library Documentation](https://polkadot.js.org/docs/)


releases
astar-3
astar-4
astar-12
astar-13
astar-14
astar-15
astar-16
astar-17
astar-18
astar-19
astar-20
astar-21
astar-22
astar-23
astar-24
astar-28
astar-30
astar-32
astar-33
astar-36
astar-37
astar-45
astar-46
dynamic-draft-test
plasm-rococo-1
plasm-rococo-2
plasm-testnet-v1
shibuya-16
shibuya-21
shibuya-22
shibuya-23
shibuya-45
shibuya-46
shibuya-50
shibuya-52
shibuya-53
shibuya-54
shibuya-57
shibuya-58
shibuya-61
shibuya-62
shibuya-63
shibuya-64
shibuya-66
shibuya-69
shibuya-70
shibuya-81
shibuya-82
shiden-1
shiden-2
shiden-2-wcp
shiden-3
shiden-6
shiden-7
shiden-8
shiden-10
shiden-14
shiden-15
shiden-16
shiden-17
shiden-18
shiden-19
shiden-20
shiden-22
shiden-23
shiden-24
shiden-31
shiden-32
shiden-33
shiden-35
shiden-36
shiden-39
shiden-42
shiden-43
shiden-44
shiden-46
shiden-47
shiden-48
shiden-49
shiden-50
shiden-51
shiden-52
shiden-53
shiden-54
shiden-55
shiden-56
shiden-57
shiden-58
shiden-59
shiden-61
shiden-64
shiden-66
shiden-67
shiden-68
shiden-69
shiden-72
shiden-73
shiden-81
shiden-82
tag
v.test
v0.0.0-evm-tracing
v0.0.0-tracing-test
v0.2.0
v0.5.0
v0.5.0pre-release
v0.6.0
v0.6.1
v0.6.2
v0.6.3
v0.7.0
v0.7.1
v0.8.0
v0.9.9
v1.0.0
v1.0.0-dusty
v1.0.0-lockdrop2
v1.0.1-dusty
v1.1.0
v1.1.0-dusty
v1.2.0-dusty
v1.2.1-dusty
v1.3.0-dusty
v1.4.0-dusty
v1.5.0-dusty
v1.5.1-dusty
v1.6.0-dusty
v1.6.1-dusty
v1.6.2-dusty
v1.7.0-dusty
v1.8.0-dusty
v1.8.1-dusty
v1.9.0-dusty
v1.9.1-dusty
v1.9.2-dusty
v1.9.3-dusty
v2.7.1
v3.0.0
v3.1.0
v3.1.1
v3.1.2
v3.1.3
v3.1.4
v3.1.6
v3.2.0
v3.2.1
v3.2.2
v3.2.3
v3.3.0
v3.4.0
v3.6.0
v3.7.1
v3.8.0
v3.9.0
v3.10.0
v3.10.1
v3.12.0
v3.14.0
v3.15.0
v3.16.0
v3.19.0
v3.19.1
v3.20.0
v3.21.0
v3.22.0
v3.23.0
v3.24.0
v3.24.1
v3.24.2
v3.25.0
v3.26.0
v3.26.1
v3.27.0
v3.28.0
v3.29.0
v3.30.0
v4.0.0
v4.1.0
v4.2.0
v4.3.0
v4.3.1
v4.4.0
v4.6.0
v4.6.1
v4.7.0
v4.9.0
v4.12.0
v4.13.0
v4.15.0
v4.17.0
v4.18.0
v4.18.1
v4.19.0
v4.21.0
v4.23.0
v4.24.0
v4.27.0
v4.29.1
v4.30.0
v4.32.0
v4.33.0
v4.34.0
v4.38.0
v4.39.0
v4.39.1
v4.41.0
v4.44.0
v4.44.0-test
v4.46.1
v4.47.1
v4.49.0
v5.0.0
v5.0.0-pre1
v5.0.0-pre2
v5.0.0-pre3
v5.0.0-pre4
v5.0.0-pre5
v5.0.0-pre6
v5.0.0-pre7
v5.0.1
v5.1.0
v5.3.0
v5.4.0
v5.9.0
v5.10.0
v5.11.0
v.test.2
v.test.3


scripts/rpc.sh
#!/bin/bash
set -e

FILE="astar-rpc"
SERVICE="astar-rpc"

function usage {
   cat << EOF
Usage: rpc.sh -auth <ngrok auth token>

Runs RPC node from docker with ngrok tunnel setting
EOF
   exit 1
}

if [ $# -ne 2 ]; then
   usage;
fi

if [ "$1" != "-auth" ]; then
  usage;
fi

AUTH=$2

# get container
sudo docker pull staketechnologies/astar-collator

# command to run RPC node - do not forget to change NAME to whatever you like
docker run -m 5G --name Shiden -p 30334:30334 -p 30333:30333 -p 9933:9933 -p 9944:9944 \
-v "/var/lib/astar/shiden-db:/data" \
-u $(id -u ${USER}):$(id -g ${USER}) -d --network=host staketechnologies/astar-collator \
astar-collator --name NAME --base-path /data --port 30333 --rpc-port 9933 --unsafe-rpc-external --unsafe-ws-external --pruning archive \
--state-cache-size 1 --telemetry-url 'wss://telemetry.polkadot.io/submit/ 0' \
-l evm=debug,ethereum=debug,rpc=debug \

sudo snap install ngrok

ngrok authtoken $AUTH
ngrok http 9933
